WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt

```diff
@@ -0,0 +1,187 @@
+import org.wordpress.android.ui.posts.editor.EditorMedia
+import org.wordpress.android.ui.posts.editor.EditorMediaListener
+import org.wordpress.android.ui.posts.editor.EditorTracker
+import org.wordpress.android.ui.posts.editor.media.OptimizeAndAddMediaToEditorUseCase.AddMediaToEditorUiState.AddingMediaIdle
+import org.wordpress.android.ui.posts.editor.media.OptimizeAndAddMediaToEditorUseCase.AddMediaToEditorUiState.AddingMultipleMedia
+import org.wordpress.android.ui.posts.editor.media.OptimizeAndAddMediaToEditorUseCase.AddMediaToEditorUiState.AddingSingleMedia
+import org.wordpress.android.ui.utils.UiString.UiStringRes
+import org.wordpress.android.util.FluxCUtilsWrapper
+import org.wordpress.android.util.MediaUtilsWrapper
+import org.wordpress.android.util.helpers.MediaFile
+import org.wordpress.android.viewmodel.SingleLiveEvent
+import javax.inject.Inject
+import javax.inject.Named
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Processes a list of media in the background (optimizing, resizing, rotating, etc.) and adds them to
+ * the editor one at a time.
+ */
+class OptimizeAndAddMediaToEditorUseCase @Inject constructor(
+    private val editorTracker: EditorTracker,
+    private val mediaUtilsWrapper: MediaUtilsWrapper,
+    private val fluxCUtilsWrapper: FluxCUtilsWrapper,
+    @Named(UI_THREAD) private val mainDispatcher: CoroutineDispatcher,
+    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher
+) : CoroutineScope {
+    private var job: Job = Job()
+
+    override val coroutineContext: CoroutineContext
+        get() = mainDispatcher + job
+
+    private val _uiState: MutableLiveData<AddMediaToEditorUiState> = MutableLiveData()
+    val uiState: LiveData<AddMediaToEditorUiState> = _uiState
+
+    private val _snackBarMessage = SingleLiveEvent<SnackbarMessageHolder>()
+    val snackBarMessage = _snackBarMessage as LiveData<SnackbarMessageHolder>
+
+    init {
+        _uiState.value = AddingMediaIdle
+    }
+
+    fun optimizeAndAddAsync(
+        uriList: List<Uri>,
+        site: SiteModel,
+        isNew: Boolean,
+        editorMediaListener: EditorMediaListener,
+        editorMedia: EditorMedia
+    ) {
+        launch {
+            _uiState.value = if (uriList.size > 1) {
+                AddingMultipleMedia
+            } else {
+                AddingSingleMedia
+            }
+
+            val optimizeMediaResult = optimizeMediaAsync(site, isNew, uriList)
+            val enqueuedFiles = enqueueMediaForUpload(
+                    optimizeMediaResult.optimizedMediaUris,
+                    editorMedia
+            )
+            addMediaToEditor(editorMediaListener, enqueuedFiles)
+
+            editorMediaListener.savePostAsyncFromEditorMedia()
+
+            if (optimizeMediaResult.someMediaCouldNotBeRetrieved) {
+                _snackBarMessage.value = SnackbarMessageHolder(R.string.gallery_error)
+            }
+
+            _uiState.value = AddingMediaIdle
```

Comment:
Similar to my comment, should this be a `postValue` call instead since it's in a bg thread?

Reply:
As explained above.